### DISCUSSION TITLE
[Feature] Custom Fields

### DISCUSSION THREAD

**Discussion by balporig** [TechbeeAT/jtxBoard#1257] [Ideas]
**Is your feature request related to a problem? Please describe.**

Sometimes you may need to define repeating common data between a collection of tasks/notes. Being able to define custom filed can open a huge set of new possibilities, making jtx journal/notes/tasks customizable beyond what the iCal standard offers

Any custom data can be stored in `description` but if you have repeatable same kind of data you have to manually input that doesn't exactly fit the description. For example, you have a shopping list and you what to define unit (g, kg, box, bottle etc.) and shop for each item and be able to select from predefined list of units and shops.

**Describe the solution you'd like**

The way I see it can work in jtx is by defining **custom fields** as a preset. Each custom field needs:
- property name
- value type: `text` (any arbitrary string like the description), `list/selection/enum/choice` (Must exactly match one or more of the prescribed choices), probably can be extended with more types in the future
- choices: list of choices if the `value type` is choice 
- collection: `collection` for which this field is available

Then for each **custom fields** preset, the field with `property name` will be available for all the journal/notes/tasks matching the `collection`.  If the value type is `text` it will be a simple text area. For `choices` it can be the same as categories or a checkbooks list. 

The commonly agreed way for extending iCal with custom properties is to prefix the property name with `X-`

```bash
X-<property name>:<value>
```

From https://www.rfc-editor.org/rfc/rfc5545

>  x-name        = "X-" [vendorid "-"] 1*(ALPHA / DIGIT / "-")
> ; Reserved for experimental use.

A blacklist for vendor specific `X-` extensions will be needed, so that they can be ignored and not show in the UI. For eg. this is what Thunderbird adds when editing tasks:

```yaml
X-MOZ-GENERATION:2
```

**Describe alternatives you've considered**

N/A

**Additional context**

The expected behavior from other iCal clients is to ignore all properties starting with `X-` but some might discard them, but IMO this is a non-issue

**Your current device and version**

N/A

**Comment by patrickunterwegs** [comment]
Hi @balporig !
Sorry for the late reply. 
Basically one of the intentions behind jtx Board was to fully support the open standard. I am aware that X-properties would allow additional properties that are not defined in the standard and basically your idea sounds tempting. 
But starting to implement a couple of X-properties that are so customized, that other apps (that don't use them) for sure wouldn't know how to interpret them, goes against my ideas of supporting the standard. It's the contrary, I would again just come up with a custom implementation of something, that is not interoperable with other apps. 

Even when I introduced markdown support for the description-field there were users who requested and appreciated it, but there were also users that complained because this would not be interoperable according to the standard. 
I am by the way using the X-properties as an extended status, so that users can define their own status as the standard is a bit too restrictive. This is especially interesting when working with the kanban-board. However, even there I took care that each extended status is mapped back to a status from the standard to always ensure that other apps can deal with it. 

If there are really convincing use-cases for this, we can for sure keep the discussion open, but for now I would not consider working on something like that. I hope you understand.

**Reply by ericfischereu** [reply]
Hi @patrickunterwegs 
I understand your arguments and yes, custom X- properties are not common standards so not every application will understand them. But in my opinion that's acceptable for this use case:

1.  Custom fields are created with the fact in mind that they are - CUSTOM. No one can expect that they will work in other applications except if they support custom fields, too.
2. X- properties are part of the standard. Usually you don't break anythink, except you take the X-MOZ properties. However, they enhance the feature set a lot.

They will practically behave like the VJOURNAL - they are standard but not supported in big CalDAV providers like Nextcloud. So jtx unfortunately already has features that others don't support out of the box.

So I think custom properties have more advantages than disadvantages.

Looking forward to your reply.
